File: projects/order-creation-queries/repository/src/main/java/com/robinfood/repository/orderdetail/OrderDetailRepository.java
package com.robinfood.repository.orderdetail;

import com.robinfood.core.dtos.OrderDetailDTO;
import com.robinfood.core.entities.APIResponseEntity;
import com.robinfood.core.entities.OrderDetailEntity;
import com.robinfood.core.enums.Result;
import com.robinfood.core.extensions.NetworkExtensionsKt;
import com.robinfood.core.mappers.OrderDetailMappers;
import java.util.List;

import com.robinfood.network.api.OrderBcAPI;
import org.springframework.stereotype.Repository;

/**
 * Implementation of IOrderDetailRepository
 */
@Repository
public class OrderDetailRepository implements IOrderDetailRepository {

    private final OrderBcAPI orderBcAPI;

    public OrderDetailRepository(OrderBcAPI orderBcAPI) {
        this.orderBcAPI = orderBcAPI;
    }

    @Override
    public Result<List<OrderDetailDTO>> getOrderDetail(
            String token,
            List<Long> orderIds,
            List<String> orderUids,
            List<String> orderUuid
    ) {
        final Result<APIResponseEntity<List<OrderDetailEntity>>> result = NetworkExtensionsKt.safeAPICall(
                orderBcAPI.getOrderDetail(
                        token,
                        orderIds,
                        orderUids,
                        orderUuid
                )
        );

        if (result instanceof Result.Error) {
            final Result.Error resultError = (Result.Error) result;
            return new Result.Error(resultError.getException(), resultError.getHttpStatus());
        }

        final Result.Success<APIResponseEntity<List<OrderDetailEntity>>> data =
                ((Result.Success<APIResponseEntity<List<OrderDetailEntity>>>) result);

        return new Result.Success(
                OrderDetailMappers.toOrderDetailDTOs(data.getData().getData())
        );
    }
}

File: projects/posv2/repository/src/main/kotlin/com/robinfood/repository/orderdetail/OrderDetailRepository.kt
package com.robinfood.repository.orderdetail

import com.robinfood.core.dtos.OrderDetailDTO
import com.robinfood.core.enums.Result
import com.robinfood.core.exceptions.OrchestratorException
import com.robinfood.core.extensions.callServices
import com.robinfood.core.extensions.safeApiCall
import com.robinfood.core.mappers.toOrderDetailDTO
import com.robinfood.network.api.OrderCreationQueriesAPI
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Repository

/**
 * Implementation of IOrderDetailRepository
 */
@Repository
class OrderDetailRepository (
        private val orderCreationQueriesAPI: OrderCreationQueriesAPI
): IOrderDetailRepository {

    override suspend fun getOrdersDetail(
            token: String,
            countryId: Long,
            flowId: Long,
            orderIds: List<Long>
    ): Result<List<OrderDetailDTO>> {

        return withContext(Dispatchers.IO) {
            val result = safeApiCall(
                    call = {
                        orderCreationQueriesAPI.getOrdersDetail(
                                token,
                                countryId,
                                flowId,
                                orderIds
                        ).callServices()
                    }
            )
            when(result) {
                is Result.Success -> {
                    val orderDetailEntities = result.data.data
                    if (orderDetailEntities == null) {
                        Result.Error(OrchestratorException("Orders detail array is null"), HttpStatus.INTERNAL_SERVER_ERROR)
                    } else {
                        val orderDetailDTOs = orderDetailEntities.mapNotNull { orderDetailEntity -> orderDetailEntity.toOrderDetailDTO() }
                        if (orderDetailDTOs.isEmpty()) {
                            Result.Error(OrchestratorException("Orders detail array is empty"), HttpStatus.INTERNAL_SERVER_ERROR)
                        } else {
                            Result.Success(orderDetailDTOs)
                        }
                    }
                }
                is Result.Error -> result
            }
        }
    }
}